public/app/plugins/datasource/prometheus/types.ts

```diff
@@ -165,3 +166,26 @@ export enum LegendFormatMode {
   Verbose = '__verbose',
   Custom = '__custom',
 }
+
+export enum PromVariableQueryType {
+  LabelNames,
+  LabelValues,
+  MetricNames,
+  VarQueryResult,
+  SeriesQuery,
+}
+
+export interface PromVariableQuery extends DataQuery {
+  query?: string;
+  expr?: string;
+  qryType?: PromVariableQueryType;
```

Comment:
@bohandley missspelled? This now in the persisted model, and we have this cryptic enum now numeric values in the persisted model (PromVariableQueryType)

Reply:
@torkelo I meant to spell `qryType` that way but I can change it if that is not correct or not clear. When a user selects the query type of label names, label values, etc, this is the `PromVariableQueryType`. We use this to identify which should be selected in the new variable editor. This could use some refactoring to make that more clear. 

All of this could be refactored though, as I originally built it to migrate to the new editor and then migrate back to the string to preserve the definitions in the variables list. I believe we are moving away from that so all of this could be much clearer with less code.

Reply:
@bohandley well this is far from clear and is now in millions of dashboard json files as this part of the saved model, so really hurts the as code story (where dashboards are defined using the save model). Not only is "qryType" a very ugly thing to have as a property type in code but also in the save model, AND it just just very cryptic numbers (0, 1, 2, 3) that have no meaning, enums must always have string values so we can make sense of them when looking at the save model.

Reply:
and why do do we have query property and expr property??  some code comments on what these properties are would also help :)

Reply:
If anyone would change the order of that enum type without a migration many dashboards would now break

Reply:
btw, think the PR as whole is awesome, just need to make sure the save model is clear and not fragile and inscrutable :)

Reply:
> @bohandley well this is far from clear and is now in millions of dashboard json files as this part of the saved model, so really hurts the as code story (where dashboards are defined using the save model). Not only is "qryType" a very ugly thing to have as a property name in code but also in the save model, AND it just very cryptic numbers (0, 1, 2, 3) that have no meaning, enums must always have string values so we can make sense of them when looking at the save model.

@torkelo thank you! This is true, I had not considered the code story as it relates to dashboards, my apologies. This makes total sense. To see `qryType: 1` in the dashboard json is both ugly and unclear as it relates to the code story. I will make sure any enums have string values moving forward. 

> and why do do we have query property and expr property?? some code comments on what these properties are would also help :)

Yes! I can add the code comments to the appropriate spot. I have some in the editor but I can see that this doesn't help when developers are reading the types to understand building a dashboard as code. The `query` and `expr` property are both there because I also had to migrate from standard variable support (which uses `expr`) and custom variable support (which uses `query`). So both fields would be present. I can add code comments for that!

https://github.com/grafana/grafana/blob/5aacdd9f49a96f4663aeb2900d4ef76ac850b98b/public/app/plugins/datasource/prometheus/components/VariableQueryEditor.tsx#L71-L73

I am sure there is a cleaner way to write all of this now, and I will create an issue to clean up some some of this. I will also add a reference point to make sure that any changes in the save model are as you say "clear, not fragile and inscrutable" 💯 Thank you for the feedback @torkelo!

Reply:
@bohandley great!  :+1:

Reply:
For the issues: https://github.com/grafana/grafana/issues/75175